File: database-driver-api/src/main/java/de/lino/database/utils/export/transcript/format/PageLayout.java
package de.lino.database.utils.export.transcript.format;

import de.lino.database.utils.export.transcript.TranscriptExporter;

import java.util.Objects;

/**
 * A page's size and orientation, e.g. A4 portrait, passed to
 * {@link TranscriptExporter#export} so a caller can choose how a transcript is laid
 * out rather than either implementation hardcoding a single page size.
 *
 * @param format the page's ISO 216 size
 * @param orientation the page's orientation
 */
public record PageLayout(PageFormat format, PageOrientation orientation) {

    /**
     * The layout every export used before {@link PageLayout} existed: A4 portrait.
     */
    public static final PageLayout DEFAULT = new PageLayout(PageFormat.A4, PageOrientation.PORTRAIT);

    public PageLayout {
        Objects.requireNonNull(format, "@PageLayout: format must not be null");
        Objects.requireNonNull(orientation, "@PageLayout: orientation must not be null");
    }

}

File: python/database-driver-api/src/database_driver/api/utils/export/transcript/format/page_layout.py
"""Mirror of ``de.lino.database.utils.export.transcript.format.PageLayout``."""

from __future__ import annotations

from dataclasses import dataclass
from typing import ClassVar

from database_driver.api.utils.export.transcript.format.page_format import PageFormat
from database_driver.api.utils.export.transcript.format.page_orientation import PageOrientation


@dataclass(frozen=True, slots=True)
class PageLayout:
    """A page's size and orientation, e.g. A4 portrait, passed to
    ``TranscriptExporter.export`` so a caller can choose how a transcript is laid out
    rather than each implementation hardcoding a single page size.

    Attributes:
        format: The page's ISO 216 size.
        orientation: The page's orientation.
    """

    format: PageFormat
    orientation: PageOrientation

    DEFAULT: ClassVar[PageLayout]
    """The layout every export used before ``PageLayout`` existed: A4 portrait."""

    def __post_init__(self) -> None:
        if not isinstance(self.format, PageFormat):
            raise TypeError("@PageLayout: format must be a PageFormat")
        if not isinstance(self.orientation, PageOrientation):
            raise TypeError("@PageLayout: orientation must be a PageOrientation")


PageLayout.DEFAULT = PageLayout(PageFormat.A4, PageOrientation.PORTRAIT)
